File: Leetcode_with_Cpp/589preorder.cpp
class Solution {
public:
    vector<int> res;
    vector<int> preorder(Node* root) {
        if(!root)   return res;
        res.push_back(root->val);
        for(auto i : root->children){
            preorder(i);
        }
        return res;
    }
};

class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> res;
        if(!root)   return res;
        stack<Node*> stk;
        stk.push(root);
        while(!stk.empty()){
            Node* tmp = stk.top();
            stk.pop();
            res.push_back(tmp->val);
            for(int i = tmp->children.size()-1; i >= 0; --i){
                stk.push(tmp->children[i]);
            }
        }
        return res;
    }
};

File: Leetcode_with_py/589preorder.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution(object):
    def preorder(self, root):
        if not root:
            return []
        res, stack = [], [root]
        while stack:
            root = stack.pop()
            res.append(root.val)
            if root.children:
                for idx in range(len(root.children)-1, -1, -1):
                    stack.append(root.children[idx])
        return res

class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        if not root:
            return []
        res = [root.val]
        if root.children:
            for chi in root.children:
                res.extend(self.preorder(chi))
        return res